File: go/packets/client/pm_test.go
package client_test

import (
	"testing"

	"github.com/goldenm-software/layrz-protocol/go/v3/packets/client"
)

func TestPm_FromPacket_ToPacket(t *testing.T) {
	tests := []struct {
		name        string
		filename    string
		contentType string
		data        []byte
	}{
		{
			name:        "text file",
			filename:    "test.txt",
			contentType: "text/plain",
			data:        []byte("hello"),
		},
		{
			name:        "binary file",
			filename:    "test.bin",
			contentType: "application/octet-stream",
			data:        []byte{0x01, 0x02, 0x03},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			packet := client.PmPacket{
				Filename:    stringPtr(tt.filename),
				ContentType: stringPtr(tt.contentType),
				Data:        &tt.data,
			}
			encoded := *packet.ToPacket()

			raw := encoded
			decoded := client.PmPacket{}
			if err := decoded.FromPacket(&raw); err != nil {
				t.Fatalf("FromPacket failed: %v", err)
			}
			if *decoded.Filename != tt.filename {
				t.Errorf("filename mismatch: got %s, want %s", *decoded.Filename, tt.filename)
			}
			if *decoded.ToPacket() != encoded {
				t.Errorf("round-trip mismatch")
			}
		})
	}
}

File: dart/test/pm_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:layrz_protocol/utils/errors.dart';
import 'package:test/test.dart';
import 'package:layrz_protocol/packets/packets.dart';
import 'package:layrz_protocol/utils/crc.dart';

void main() {
  test('Packet.fromPacket() routes PmPacket', () {
    final original = PmPacket(
      filename: 'test.bin',
      contentType: 'application/octet-stream',
      data: Uint8List.fromList([0x01, 0x02]),
    );
    final parsed = Packet.fromPacket(original.toPacket());
    expect(parsed, isA<PmPacket>());
  });

  test('PmPacket.toPacket()', () {
    final data = Uint8List.fromList([0x48, 0x65, 0x6C, 0x6C, 0x6F]); // "Hello"
    final packet = PmPacket(filename: 'test.txt', contentType: 'text/plain', data: data);
    final raw = packet.toPacket();

    expect(raw.startsWith('<Pm>'), true);
    expect(raw.endsWith('</Pm>'), true);

    String b64 = base64Encode(data);
    String payload = 'test.txt;text/plain;$b64;';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    expect(raw, '<Pm>$payload$crc</Pm>');
  });

  test('PmPacket.fromPacket()', () {
    final data = Uint8List.fromList([0x48, 0x65, 0x6C, 0x6C, 0x6F]);
    String b64 = base64Encode(data);
    String payload = 'test.txt;text/plain;$b64;';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    String raw = '<Pm>$payload$crc</Pm>';

    final packet = PmPacket.fromPacket(raw);
    expect(packet.filename, 'test.txt');
    expect(packet.contentType, 'text/plain');
    expect(packet.data, data);
    expect(packet.toPacket(), raw);
  });

  test('PmPacket.fromPacket() invalid format throws ParseException', () {
    expect(() => PmPacket.fromPacket('invalid'), throwsA(isA<ParseException>()));
  });

  test('PmPacket.fromPacket() wrong part count throws MalformedException', () {
    String payload = 'only;two;';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();
    expect(() => PmPacket.fromPacket('<Pm>$payload$crc</Pm>'), throwsA(isA<MalformedException>()));
  });

  test('PmPacket.copyWith()', () {
    final data = Uint8List.fromList([0x01]);
    final original = PmPacket(filename: 'a.bin', contentType: 'application/octet-stream', data: data);
    final copy = original.copyWith(filename: 'b.bin');
    expect(copy.filename, 'b.bin');
    expect(copy.contentType, 'application/octet-stream');
    expect(copy.data, data);
  });
}
